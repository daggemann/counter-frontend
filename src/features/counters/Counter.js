import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL + '/counters';

export class Counter {

    constructor(id, title, count) {
        this._id = id;
        this._title = title;
        this._count = count;
    }

    get id() {
        return this._id;
    }

    set id(id) {
        this._id = id;
    }

    get title() {
        return this._title;
    }

    set title(title) {
        this._title = title;
    }

    get count() {
        return this._count;
    }

    set count(count) {
        this._count = count;
    }

    increment() {
        return new Promise( (resolve, reject) => {
            axios.post(API_URL + `/${this._id}/increment`, {})
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }

    delete() {
        return new Promise( (resolve, reject) => {
            axios.delete(API_URL + `/${this._id}`)
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }

    decrement() {
        return new Promise( (resolve, reject) => {
            axios.post(API_URL + `/${this._id}/decrement`, {})
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }
}


export function totalCount(counters) {
    let totalCount = 0;
    counters.forEach((counter) => {
        totalCount += counter.count;
    });
    return totalCount;
}

export function getAll() {
    return new Promise((resolve, reject) => {
        axios.get(API_URL)
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })

    })
}

export function add(title) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL, {title: title})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function incrementCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL + `/${id}/increment`, {})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function decrementCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL + `/${id}/decrement`, {})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function deleteCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.delete(API_URL + `/${id}`)
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}
