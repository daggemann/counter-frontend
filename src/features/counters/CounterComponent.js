import React from 'react';
import styled from 'styled-components';
import {Button, FlexRow} from "../../commons/styled-components";

const DeleteButton = Button.extend`
    &:hover {
        margin-right: 10px;
    }
`;
DeleteButton.displayName = 'DeleteButton';

const Title = styled.span`
    margin-top: 6px;
    text-align: center;
`;
Title.displayName = 'Title';

const DecrementButton = Button.extend`
    &:hover {
        background-color: #7FDBFF;
    }

`;
DecrementButton.displayName = 'DecrementButton';

const Count = styled.span`
    text-align: center;
    margin-top: 6px;
    width: 50px;
`;
Count.displayName = 'Count';

const IncrementButton = Button.extend`
    &:hover {
        background-color: #7FDBFF;
    }
`;
IncrementButton.displayName = 'IncrementButton';

export default function CounterComponent(props) {
    return (
        <FlexRow justifyContent='space-between' margin='5px 0px 0px 0px' height='30px'>
            <FlexRow>
                <DeleteButton
                    onClick={() => props.delete(props.counter.id)}
                    width='30px'
                    margin='0px 2px 0px 0px'
                    border='0px'
                    borderRadius='1em 0em 0em 1em'
                    backgroundColor='#FF4136'
                >
                    x
                </DeleteButton>
                <Title>{props.counter.title}</Title>
            </FlexRow>
            <FlexRow width='100px'>
                <DecrementButton
                    onClick={() => props.decr(props.counter.id)}
                    width='20px'
                    border='0px'
                    borderRadius='1em 0em 0em 1em'
                >
                    -
                </DecrementButton>
                <Count>{props.counter.count}</Count>
                <IncrementButton
                    onClick={() => props.incr(props.counter.id)}
                    width='20px'
                    border='0px'
                    borderRadius='0em 1em 1em 0em'
                >
                    +
                </IncrementButton>
            </FlexRow>
        </FlexRow>
    )
}