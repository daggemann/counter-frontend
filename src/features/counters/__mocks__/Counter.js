export class Counter {
    constructor(id, title, count) {
        this.id = id;
        this.title = title;
        this.count = count;
    }

    delete() {
        delete counters[this.id]
    }
}

let counters = [
    new Counter(1, "test", 1),
    new Counter(2, "test", 1),
    new Counter(3, "test", 1),
];


export function getAll() {
    return new Promise((resolve, reject) => {
        (true) ? setTimeout(resolve(counters.slice()), 3) : reject('never gonna happen')
    });
}

export function add(title) {
    counters.push(new Counter(4, "test", 1));
    return new Promise((resolve, reject) => {
        (true) ? setTimeout(resolve(counters.slice()), 3) : reject('never gonna happen')
    });
}

export function totalCount(counters) {
    let totalCount = 0;
    counters.forEach((counter) => {
        totalCount += counter.count;
    });
    return totalCount;
}

