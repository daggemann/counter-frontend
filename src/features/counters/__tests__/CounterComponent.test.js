import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import CounterComponent from "../CounterComponent";
import {Counter} from '../Counter';

test('render counter', () => {
    const counter = new Counter(1, "test", 1);
    const tree = renderer.create(<CounterComponent counter={counter}/>);
    expect(tree.toJSON()).toMatchSnapshot();
});

test('ensure that title and count are displayed', () => {
    const counter = new Counter(1, "test", 1);
    const component = shallow(<CounterComponent counter={counter}/>);
    const title = component.find('Title');
    const count = component.find('Count');
    expect(title.get(0).props.children).toEqual("test");
    expect(count.get(0).props.children).toEqual(1);
});

test('delete counter', () => {
    const mockDelete = jest.fn();
    const counter = new Counter(1, "test", 1);
    const component = shallow(<CounterComponent delete={(i) => mockDelete(i)} counter={counter}/>);
    const deleteButton = component.find('DeleteButton');
    deleteButton.simulate('click');
    expect(mockDelete).toHaveBeenCalledTimes(1);
    expect(mockDelete).toBeCalledWith(1)
});

test('increment counter', () => {
    const mockIncrement = jest.fn();

    const counter = new Counter(1, "test", 1);
    const component = shallow(<CounterComponent incr={(i) => mockIncrement(i)} counter={counter}/>);
    const incrementButton = component.find('IncrementButton');
    incrementButton.simulate('click');
    expect(mockIncrement).toHaveBeenCalledTimes(1);
    expect(mockIncrement).toBeCalledWith(1);
});

test('decrement counter', () => {
    const mockDecrement = jest.fn();

    const counter = new Counter(1, "test", 1);
    const component = shallow(<CounterComponent decr={(i) => mockDecrement(i)} counter={counter}/>);
    const incrementButton = component.find('DecrementButton');
    incrementButton.simulate('click');
    expect(mockDecrement).toHaveBeenCalledTimes(1);
    expect(mockDecrement).toBeCalledWith(1);
});

