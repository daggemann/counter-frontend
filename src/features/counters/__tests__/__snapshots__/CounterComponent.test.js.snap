// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render counter 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 5px 0px 0px 0px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100px;
}

.c2 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 30px;
  background-color: #FF4136;
  margin: 0px 2px 0px 0px;
}

.c2:hover {
  margin-right: 10px;
}

.c3 {
  margin-top: 6px;
  text-align: center;
}

.c5 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 20px;
}

.c5:hover {
  background-color: #7FDBFF;
}

.c6 {
  text-align: center;
  margin-top: 6px;
  width: 50px;
}

.c7 {
  border: 0px;
  border-radius: 0em 1em 1em 0em;
  width: 20px;
}

.c7:hover {
  background-color: #7FDBFF;
}

<div
  className="c0"
  height="30px"
>
  <div
    className="c1"
  >
    <button
      className="c2"
      onClick={[Function]}
      width="30px"
    >
      x
    </button>
    <span
      className="c3"
    >
      test
    </span>
  </div>
  <div
    className="c4"
    width="100px"
  >
    <button
      className="c5"
      onClick={[Function]}
      width="20px"
    >
      -
    </button>
    <span
      className="c6"
    >
      1
    </span>
    <button
      className="c7"
      onClick={[Function]}
      width="20px"
    >
      +
    </button>
  </div>
</div>
`;
